cadc-util/src/main/java/ca/nrc/cadc/auth/NotAuthenticatedException.java

```diff
@@ -78,12 +78,64 @@
  */
 public class NotAuthenticatedException extends SecurityException {
     
+    private String challenge;
+    private OAuthError oauthError;
+    
```

Comment:
Aside from the fact that VO people will quibble about the actual values, I think this could be usable with ivoa challenge failures as well... as noted in RestServlet we should report failed authentication for ivoa challenge. And we will have expired ivoa tokens :-)

In theory, this could occur with client cert if the Authenticator.validate(...) impl imposed further restrictions not enforced by the CA bundle.

Suggestion:  maybe rename the things in code (enum, arg names) to just AuthError and have a comment that the words are from the OAuth2 API... just cosmetic really: the important thing is that our validate adds the OAuthError when throwing the exception with an ivoa challenge.

Reply:
Yes, I think those names are misleading, I'll remove the OAuth.  I think (at least I intended) to make that error handling work with both types of tokens.  If you look in TokenValidator where the exceptions are constructed it doesn't differentiate.